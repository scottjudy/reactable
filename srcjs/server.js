import React from 'react'
import * as ReactDOMServer from 'react-dom/server'
import createEmotionServer from '@emotion/server/create-instance'

import Reactable from './Reactable'
import { evaluateStringMember } from './htmlwidgets'
import './intl-polyfill'
import { getEmotion } from './theme'

// Provide React as a global for users that call React externally in custom JS render functions.
// Although `global` may not necessarily be available in V8, it is in reactable's V8 context,
// and Webpack transforms this to a portable assignment that doesn't assume `global` exists anyway.
global.React = React

const cache = getEmotion().cache
const { extractCritical } = createEmotionServer(cache)

export function renderToHTML(inputString) {
  const input = JSON.parse(inputString)

  const props = input.props
  props.data = JSON.parse(props.data)

  // Resolve strings marked as JavaScript literals to objects
  if (input.evals) {
    for (let key of input.evals) {
      evaluateStringMember(props, key)
    }
  }

  const { html, css, ids } = extractCritical(ReactDOMServer.renderToString(<Reactable {...props} />))
  return { html, css, ids }
}
