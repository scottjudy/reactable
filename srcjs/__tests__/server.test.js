/**
 * @jest-environment node
 */

import { renderToHTML } from '../server'

// These tests run in Node and are intentionally minimal. See test-reactable.R for
// more comprehensive tests that run in true V8 environments (different from Node).

describe('renderToHTML', () => {
  it('renders tables to HTML', () => {
    const inputString =
      '{"props":{"data":"{\\"Manufacturer\\":[\\"Acura\\",\\"Acura\\",\\"Audi\\"],\\"Model\\":[\\"Integra\\",\\"Legend\\",\\"90\\"],\\"Type\\":[\\"Small\\",\\"Midsize\\",\\"Compact\\"]}","columns":[{"id":"Manufacturer","name":"Manufacturer","type":"factor"},{"id":"Model","name":"Model","type":"factor"},{"id":"Type","name":"Type","type":"factor"}],"dataKey":"b540d0d929ae2ade6248d0cbb7f477fc"},"evals":[]}'
    const { html, css, ids } = renderToHTML(inputString)
    expect(html).toMatchSnapshot()
    expect(css).toEqual('')
    expect(ids).toEqual([])
  })

  it('renders tables to HTML with JS evals', () => {
    const inputString =
      '{"props":{"data":"{\\"Manufacturer\\":[\\"Acura\\",\\"Acura\\"],\\"Model\\":[\\"Integra\\",\\"Legend\\"]}","columns":[{"id":"Manufacturer","name":"Manufacturer","type":"factor"},{"id":"Model","name":"Model","type":"factor","cell":"cellInfo => cellInfo.value + \'__\'"}],"dataKey":"7951e598fecb5fa57c37a854a0917a7c"},"evals":["columns.1.cell"]}'
    const { html, css, ids } = renderToHTML(inputString)
    expect(html).toContain('Integra__')
    expect(html).toContain('Legend__')
    expect(html).toMatchSnapshot()
    expect(css).toEqual('')
    expect(ids).toEqual([])
  })

  it('renders tables to HTML with theme styles', () => {
    const inputString =
      '{"props":{"data":"{\\"Manufacturer\\":[\\"Acura\\",\\"Acura\\"],\\"Model\\":[\\"Integra\\",\\"Legend\\"]}","columns":[{"id":"Manufacturer","name":"Manufacturer","type":"factor"},{"id":"Model","name":"Model","type":"factor"}],"theme":{"color":"red","cellPadding":"1rem"},"dataKey":"01768d44dd8153352ff1cfa908f38d72"},"evals":[]}'
    const { html, css, ids } = renderToHTML(inputString)
    expect(html).toMatchSnapshot()
    expect(css).toEqual('.reactable-1galy0v{color:red;}.reactable-1rsvkai{padding:1rem;}')
    expect(ids).toEqual(['1galy0v', '1xdhyk6', '1rsvkai'])
  })
})
