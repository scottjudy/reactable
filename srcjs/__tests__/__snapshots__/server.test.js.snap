// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderToHTML renders tables to HTML 1`] = `"<div class=\\"Reactable ReactTable\\"><div class=\\"rt-table\\" role=\\"table\\"><div class=\\"rt-thead\\" role=\\"rowgroup\\" style=\\"min-width:300px\\"><div class=\\"rt-tr rt-tr-header\\" role=\\"row\\" style=\\"flex:1 0 auto;min-width:300px\\"><div class=\\"rt-th rt-align-left\\" role=\\"columnheader\\" style=\\"position:relative;flex:100 0 auto;min-width:100px;width:100px\\" aria-label=\\"Sort Manufacturer\\" aria-sort=\\"none\\" tabindex=\\"0\\" data-sort-hint=\\"ascending\\"><div class=\\"rt-th-inner\\"><div class=\\"rt-sort-header\\"><div class=\\"rt-text-content\\">Manufacturer</div><span class=\\"rt-sort-right\\" aria-hidden=\\"true\\"></span></div></div></div><div class=\\"rt-th rt-align-left\\" role=\\"columnheader\\" style=\\"position:relative;flex:100 0 auto;min-width:100px;width:100px\\" aria-label=\\"Sort Model\\" aria-sort=\\"none\\" tabindex=\\"0\\" data-sort-hint=\\"ascending\\"><div class=\\"rt-th-inner\\"><div class=\\"rt-sort-header\\"><div class=\\"rt-text-content\\">Model</div><span class=\\"rt-sort-right\\" aria-hidden=\\"true\\"></span></div></div></div><div class=\\"rt-th rt-align-left\\" role=\\"columnheader\\" style=\\"position:relative;flex:100 0 auto;min-width:100px;width:100px\\" aria-label=\\"Sort Type\\" aria-sort=\\"none\\" tabindex=\\"0\\" data-sort-hint=\\"ascending\\"><div class=\\"rt-th-inner\\"><div class=\\"rt-sort-header\\"><div class=\\"rt-text-content\\">Type</div><span class=\\"rt-sort-right\\" aria-hidden=\\"true\\"></span></div></div></div></div></div><div class=\\"rt-tbody\\" role=\\"rowgroup\\" style=\\"min-width:300px\\"><div class=\\"rt-tr-group\\"><div class=\\"rt-tr\\" role=\\"row\\" style=\\"flex:1 0 auto;min-width:300px\\"><div class=\\"rt-td rt-align-left\\" role=\\"cell\\" style=\\"flex:100 0 auto;min-width:100px;width:100px\\"><div class=\\"rt-td-inner\\">Acura</div></div><div class=\\"rt-td rt-align-left\\" role=\\"cell\\" style=\\"flex:100 0 auto;min-width:100px;width:100px\\"><div class=\\"rt-td-inner\\">Integra</div></div><div class=\\"rt-td rt-align-left\\" role=\\"cell\\" style=\\"flex:100 0 auto;min-width:100px;width:100px\\"><div class=\\"rt-td-inner\\">Small</div></div></div></div><div class=\\"rt-tr-group\\"><div class=\\"rt-tr\\" role=\\"row\\" style=\\"flex:1 0 auto;min-width:300px\\"><div class=\\"rt-td rt-align-left\\" role=\\"cell\\" style=\\"flex:100 0 auto;min-width:100px;width:100px\\"><div class=\\"rt-td-inner\\">Acura</div></div><div class=\\"rt-td rt-align-left\\" role=\\"cell\\" style=\\"flex:100 0 auto;min-width:100px;width:100px\\"><div class=\\"rt-td-inner\\">Legend</div></div><div class=\\"rt-td rt-align-left\\" role=\\"cell\\" style=\\"flex:100 0 auto;min-width:100px;width:100px\\"><div class=\\"rt-td-inner\\">Midsize</div></div></div></div><div class=\\"rt-tr-group\\"><div class=\\"rt-tr\\" role=\\"row\\" style=\\"flex:1 0 auto;min-width:300px\\"><div class=\\"rt-td rt-align-left\\" role=\\"cell\\" style=\\"flex:100 0 auto;min-width:100px;width:100px\\"><div class=\\"rt-td-inner\\">Audi</div></div><div class=\\"rt-td rt-align-left\\" role=\\"cell\\" style=\\"flex:100 0 auto;min-width:100px;width:100px\\"><div class=\\"rt-td-inner\\">90</div></div><div class=\\"rt-td rt-align-left\\" role=\\"cell\\" style=\\"flex:100 0 auto;min-width:100px;width:100px\\"><div class=\\"rt-td-inner\\">Compact</div></div></div></div><div class=\\"rt-no-data\\" aria-live=\\"assertive\\"></div></div></div></div>"`;

exports[`renderToHTML renders tables to HTML with JS evals 1`] = `"<div class=\\"Reactable ReactTable\\"><div class=\\"rt-table\\" role=\\"table\\"><div class=\\"rt-thead\\" role=\\"rowgroup\\" style=\\"min-width:200px\\"><div class=\\"rt-tr rt-tr-header\\" role=\\"row\\" style=\\"flex:1 0 auto;min-width:200px\\"><div class=\\"rt-th rt-align-left\\" role=\\"columnheader\\" style=\\"position:relative;flex:100 0 auto;min-width:100px;width:100px\\" aria-label=\\"Sort Manufacturer\\" aria-sort=\\"none\\" tabindex=\\"0\\" data-sort-hint=\\"ascending\\"><div class=\\"rt-th-inner\\"><div class=\\"rt-sort-header\\"><div class=\\"rt-text-content\\">Manufacturer</div><span class=\\"rt-sort-right\\" aria-hidden=\\"true\\"></span></div></div></div><div class=\\"rt-th rt-align-left\\" role=\\"columnheader\\" style=\\"position:relative;flex:100 0 auto;min-width:100px;width:100px\\" aria-label=\\"Sort Model\\" aria-sort=\\"none\\" tabindex=\\"0\\" data-sort-hint=\\"ascending\\"><div class=\\"rt-th-inner\\"><div class=\\"rt-sort-header\\"><div class=\\"rt-text-content\\">Model</div><span class=\\"rt-sort-right\\" aria-hidden=\\"true\\"></span></div></div></div></div></div><div class=\\"rt-tbody\\" role=\\"rowgroup\\" style=\\"min-width:200px\\"><div class=\\"rt-tr-group\\"><div class=\\"rt-tr\\" role=\\"row\\" style=\\"flex:1 0 auto;min-width:200px\\"><div class=\\"rt-td rt-align-left\\" role=\\"cell\\" style=\\"flex:100 0 auto;min-width:100px;width:100px\\"><div class=\\"rt-td-inner\\">Acura</div></div><div class=\\"rt-td rt-align-left\\" role=\\"cell\\" style=\\"flex:100 0 auto;min-width:100px;width:100px\\"><div class=\\"rt-td-inner\\">Integra__</div></div></div></div><div class=\\"rt-tr-group\\"><div class=\\"rt-tr\\" role=\\"row\\" style=\\"flex:1 0 auto;min-width:200px\\"><div class=\\"rt-td rt-align-left\\" role=\\"cell\\" style=\\"flex:100 0 auto;min-width:100px;width:100px\\"><div class=\\"rt-td-inner\\">Acura</div></div><div class=\\"rt-td rt-align-left\\" role=\\"cell\\" style=\\"flex:100 0 auto;min-width:100px;width:100px\\"><div class=\\"rt-td-inner\\">Legend__</div></div></div></div><div class=\\"rt-no-data\\" aria-live=\\"assertive\\"></div></div></div></div>"`;

exports[`renderToHTML renders tables to HTML with theme styles 1`] = `"<div class=\\"Reactable ReactTable reactable-1galy0v\\"><div class=\\"rt-table\\" role=\\"table\\"><div class=\\"rt-thead\\" role=\\"rowgroup\\" style=\\"min-width:200px\\"><div class=\\"rt-tr rt-tr-header\\" role=\\"row\\" style=\\"flex:1 0 auto;min-width:200px\\"><div class=\\"rt-th rt-align-left reactable-1xdhyk6\\" role=\\"columnheader\\" style=\\"position:relative;flex:100 0 auto;min-width:100px;width:100px\\" aria-label=\\"Sort Manufacturer\\" aria-sort=\\"none\\" tabindex=\\"0\\" data-sort-hint=\\"ascending\\"><div class=\\"rt-th-inner reactable-1rsvkai\\"><div class=\\"rt-sort-header\\"><div class=\\"rt-text-content\\">Manufacturer</div><span class=\\"rt-sort-right\\" aria-hidden=\\"true\\"></span></div></div></div><div class=\\"rt-th rt-align-left reactable-1xdhyk6\\" role=\\"columnheader\\" style=\\"position:relative;flex:100 0 auto;min-width:100px;width:100px\\" aria-label=\\"Sort Model\\" aria-sort=\\"none\\" tabindex=\\"0\\" data-sort-hint=\\"ascending\\"><div class=\\"rt-th-inner reactable-1rsvkai\\"><div class=\\"rt-sort-header\\"><div class=\\"rt-text-content\\">Model</div><span class=\\"rt-sort-right\\" aria-hidden=\\"true\\"></span></div></div></div></div></div><div class=\\"rt-tbody\\" role=\\"rowgroup\\" style=\\"min-width:200px\\"><div class=\\"rt-tr-group\\"><div class=\\"rt-tr\\" role=\\"row\\" style=\\"flex:1 0 auto;min-width:200px\\"><div class=\\"rt-td rt-align-left reactable-1xdhyk6\\" role=\\"cell\\" style=\\"flex:100 0 auto;min-width:100px;width:100px\\"><div class=\\"rt-td-inner reactable-1rsvkai\\">Acura</div></div><div class=\\"rt-td rt-align-left reactable-1xdhyk6\\" role=\\"cell\\" style=\\"flex:100 0 auto;min-width:100px;width:100px\\"><div class=\\"rt-td-inner reactable-1rsvkai\\">Integra</div></div></div></div><div class=\\"rt-tr-group\\"><div class=\\"rt-tr\\" role=\\"row\\" style=\\"flex:1 0 auto;min-width:200px\\"><div class=\\"rt-td rt-align-left reactable-1xdhyk6\\" role=\\"cell\\" style=\\"flex:100 0 auto;min-width:100px;width:100px\\"><div class=\\"rt-td-inner reactable-1rsvkai\\">Acura</div></div><div class=\\"rt-td rt-align-left reactable-1xdhyk6\\" role=\\"cell\\" style=\\"flex:100 0 auto;min-width:100px;width:100px\\"><div class=\\"rt-td-inner reactable-1rsvkai\\">Legend</div></div></div></div><div class=\\"rt-no-data\\" aria-live=\\"assertive\\"></div></div></div></div>"`;
